### ISSUE TITLE
High CPU usage

### ISSUE THREAD

**Issue by mamantoha** [mosquito-cr/mosquito#126]
Hi @robacarp .
There is an issue that arose after commit 51904a05674757410268d183dd78fb2259ddbad7 on the shards.info production. The CPU usage remains consistently at 100%. Here's a screenshot from DigitalOcean for reference:

![image](https://github.com/mosquito-cr/mosquito/assets/61285/f9d0bf54-af80-4b19-964a-38f6aac80dab)

Upon reverting this commit, everything returns to normal:

![image](https://github.com/mosquito-cr/mosquito/assets/61285/9a6b904b-8163-4a76-9c19-2166a4600c7a)
![image](https://github.com/mosquito-cr/mosquito/assets/61285/7de26434-78e9-4191-b64c-25e3f5de26e8)

I apologize for not being able to provide more details at this moment.

**Comment by robacarp** [comment]
@mamantoha Thanks for trying out the new executor parallel feature! 

I'd be happy to help you sort this out. I've run shards.info locally with that commit and had no issue, but it doesn't mean something isn't broken. I think the log configuration for shards.info puts mosquito logs in a separate file -- can you share the logs from the worker when it's pegging the CPU to 100%?

**Comment by robacarp** [comment]
It’s possible that 100% CPU usage is actually Mosquito doing what it was designed to do. How long did the CPU burst last? I believe I’m able to do a full update scrape in just a few minutes on my laptop. 

Shards.info is a network bound workflow so running multiple threads will allow the scrape to be parallelized and scrape much faster than single threaded. 

I haven’t set a configuration variable for it yet but you can [monkey patch](https://github.com/mosquito-cr/mosquito/blob/51904a05674757410268d183dd78fb2259ddbad7/src/mosquito/runners/overseer.cr#L33) to change the number of executors which get spawned.

**Comment by mamantoha** [comment]
The CPU usage has consistently remained at 100% (Jan 3 - Jan 11), and it returns to normal after reverting to the previous mosquito commit.

![image](https://github.com/mosquito-cr/mosquito/assets/61285/51d2ad08-bfe0-400c-b53f-9c9346b6f706)

 I will try to reduce the issue on my Linux laptop over the weekend.

**Comment by robacarp** [comment]
Ah yeah, I didn't think to look at the graph. If you can send me a database dump I'm happy to run it locally and see where the problem is too.

**Comment by mamantoha** [comment]
@robacarp I successfully replicated the issue on my Linux laptop. Where can I send you a database dump?

Setting `getter executor_count = 1` didn't help.

**Comment by robacarp** [comment]
Oh interesting, thank you for the context. I'll shell into a linux VM somewhere and see if I can replicate.

I doubt the database is important.

**Comment by robacarp** [comment]
@mamantoha I've sorted it out. I spent way too much time assuming the problem was in the new executor architecture but that is completely unrelated. The [issue is here](https://github.com/mosquito-cr/mosquito/blame/51904a05674757410268d183dd78fb2259ddbad7/src/mosquito/runner.cr#L47), this line:

```diff
    def self.start(spin = true)
      Log.notice { "Mosquito is buzzing..." }
      instance.run

      while spin && keep_running
+        Fiber.yield
      end
    end
```

I made the mistake of assuming that Fiber.yield was actually communicating to the scheduler that it should transfer control to another fiber, but [that is apparently not the case](https://github.com/crystal-lang/crystal/issues/9128). I'll make a `sleep 2` patch shortly, and put in an issue for a smarter Runner one-shot class that doesn't just sleep.

Sorry for the delay in sorting this out!
